added pythonnet check with warning output

klmcadams opened the pull request:
If pythonnet is installed and a remote instance is used, there should not be a warning. Pymechanical embedding should work, but some of the APIs will not work if pythonnet is installed. Since the remote instance should still work whether or not pythonnet is installed, it was determined a warning would be a better use, so users are aware of the conflict, but it doesn't stop it from running.

RobPasMue commented on src/ansys/mechanical/core/embedding/app.py:
```diff
@@ -71,6 +74,19 @@ def __init__(self, db_file=None, **kwargs):
             self._version = _get_default_version()
         initializer.initialize(self._version)
+
+        pkgs = freeze.freeze()
+        for pkg in pkgs:
+            indx = pkg.index("=")
+            if pkg[:indx] == "pythonnet":
+                warnings.warn(
+                    "The pythonnet package was found in your environment"
+                    "which interferes with the ansys-pythonnet package. "
+                    "Some APIs may not work due to pythonnet being installed.",
+                    stacklevel=3,
+                )
+                break
```
Nice implementation @klmcadams =) could you unit test this feature if possible?

klmcadams replied:
Thanks! Sure, I will add them in soon.

RobPasMue commented on src/ansys/mechanical/core/embedding/app.py:
```diff
@@ -2,6 +2,9 @@
 import typing
+import warnings
+
+from pip._internal.operations import freeze
```
I'm not a big fan of using "private" modules, but it's fine. However it looks like at older versions of pip this module is located in a different location:
```suggestion
try:
    from pip._internal.operations import freeze
except ImportError: # pip < 10.0
    from pip.operations import freeze
```

RobPasMue commented on src/ansys/mechanical/core/embedding/app.py:
```diff
@@ -71,6 +74,19 @@ def __init__(self, db_file=None, **kwargs):
             self._version = _get_default_version()
         initializer.initialize(self._version)
+
+        pkgs = freeze.freeze()
```
Add some inline comment saying the reasons for having this code snippet
```suggestion
        # Check if 'pythonnet' is installed... and if so, throw warning
        pkgs = freeze.freeze()
```

RobPasMue reviewed (commented):
Looking good. @klmcadams please add some unit tests for everifying the correct behavior of this implementation.

Also, the documentation warnings that rise are also appearing in the main branch when the workflow is ran. Seems to me like some kind of new side-effect... See https://github.com/pyansys/pymechanical/actions/runs/4879631252/jobs/8706395546

I tested it locally and it seems like the root cause is the new Python dev docs. They must have done some change to the way they define booleans for Python 3.12.X...

Easy fix would be to change in ``conf.py`` line 63 as follows:
```"python": ("https://docs.python.org/3", None),```

Could you also take care of that?

RobPasMue commented:
@klmcadams please implement the docs fix in a different PR - just to keep them separated - and then update your branch. Thanks for your time and efforts!

klmcadams commented:
> @klmcadams please implement the docs fix in a different PR - just to keep them separated - and then update your branch. Thanks for your time and efforts!

I made a branch called fix/docwarning for the doc pipeline issue. No problem! Let me know if I can do anything else.

MaxJPRey reviewed (approved):
LGTM. Nice work.
Agreed with Roberto regarding the unit tests.

koubaa commented:
@klmcadams nice work! I'll review in detail tomorrow. Could you please update the changelog?

koubaa commented on src/ansys/mechanical/core/embedding/app.py:
```diff
@@ -71,6 +77,20 @@ def __init__(self, db_file=None, **kwargs):
             self._version = _get_default_version()
         initializer.initialize(self._version)
+
+        # Check if 'pythonnet' is installed... and if so, throw warning
```
@klmcadams would importlib-metadata work here?

koubaa commented on src/ansys/mechanical/core/embedding/app.py:
```diff
@@ -71,6 +77,20 @@ def __init__(self, db_file=None, **kwargs):
         if self._version == None:
             self._version = _get_default_version()
         initializer.initialize(self._version)
+
```
better to do this inside of `initializer.initialize` then in app.py

klmcadams commented:
> LGTM. Nice work. Agreed with Roberto regarding the unit tests.

Thank you! I will upload the unit tests as soon as possible.

klmcadams commented:
> @klmcadams nice work! I left a few comments/suggestions but I'll review in more detail tomorrow. Could you please update the changelog?

Thank you! That sounds good. I can update the changelog as well.

klmcadams commented:
Hi @koubaa, @MaxJPRey & @RobPasMue, I added the unit tests for the pythonnet changes and added a line to the changelog. 

I looked into importlib.metadata as Mohamed suggested, and think it works better than the pip freeze implementation I did earlier. Instead of parsing through the pip freeze output, it now calls the distribution('pythonnet'). If the pythonnet is in the environment, it creates the warning message, otherwise it prints a message saying pythonnet was not found. 

I added apt install -y python3.8-venv to the ci_cd.yml file because when I was getting venv errors for the embedded/launcher tests, the output said to run "apt install -y python3.8-venv" no matter what version it was. 

After checking in my most recent commit and the automatic PR was running, the Launch Testing and Coverage (3.10) step was the only one that seemed to get stuck in unit testing. Whenever I canceled that step and ran it again, it passed. I'm not sure how often this happens, but we could deploy the workflow again before merging into main just to be sure if you would like. 

I believe those are all the major updates with this PR. Let me know if you have any questions or concerns. Thanks!

RobPasMue commented on .github/workflows/ci_cd.yml:
```diff
@@ -158,6 +158,7 @@ jobs:
           apt update
           apt install -y lsb-release
           apt install -y python3-pip
+          apt install -y python3.8-venv
```
Yep this is needed for creating venvs.

RobPasMue commented on src/ansys/mechanical/core/embedding/app.py:
```diff
@@ -71,6 +71,7 @@ def __init__(self, db_file=None, **kwargs):
         if self._version == None:
             self._version = _get_default_version()
         initializer.initialize(self._version)
+
```
Since there are no real changes here, let's remove the blank line
```suggestion
```

RobPasMue commented on src/ansys/mechanical/core/embedding/initializer.py:
```diff
@@ -38,6 +40,18 @@ def initialize(version):
+            stacklevel=2,
+        )
+    except ModuleNotFoundError:
+        print("pythonnet was not found")
```
If it's not found (which is the adequate situation) I'd just pass.
```suggestion
        pass
```

RobPasMue commented on tests/embedding/test_app.py:
```diff
@@ -52,3 +54,67 @@ def test_app_version(embedded_app):
+    subprocess.run([sys.executable, "-m", "venv", "." + venv_name])
+
+    # Upgrade pip
+    print("Upgrading pip")
```
I'd personally  remove the ``print`` statements. They are fine for local dev - but do not make them part of the PR.

klmcadams replied:
Sure, that makes sense! I'll remove them

RobPasMue commented on tests/embedding/test_app.py:
```diff
@@ -52,3 +54,67 @@ def test_app_version(embedded_app):
+
+    """If UserWarning & pythonnet are in the stderr output,
+    set warning to 1. Otherwise, set warning to 0"""
+    if "UserWarning" and "pythonnet" in stderr_output:
+        warning = 1
+    else:
+        warning = 0
```
You could also do this:
```suggestion
    warning = True if "UserWarning" and "pythonnet" in stderr_output else False
```

RobPasMue commented on tests/embedding/test_app.py:
```diff
@@ -52,3 +54,67 @@ def test_app_version(embedded_app):
+        warning = 0
+
+    # Assert warning message appears for embedded app
+    assert 1 == warning
```
And then this assert would be clearer
```suggestion
    assert warning
```

RobPasMue commented on tests/test_mechanical.py:
```diff
@@ -510,6 +513,93 @@ def test_launch_grpc_not_supported_version():
+
+    """If UserWarning & pythonnet are in the stderr output,
+    set warning to 0. Otherwise, set warning to 1"""
+    if "UserWarning" and "pythonnet" in stderr_output:
+        warning = 0
+    else:
+        warning = 1
+
+    # Assert the warning message did not appear for the remote session
+    assert 1 == warning
```
```suggestion
    warning = True if "UserWarning" and "pythonnet" in stderr_output else False

    # Assert the warning message did not appear for the remote session
    assert not warning
```

RobPasMue commented on tests/test_mechanical.py:
```diff
@@ -510,6 +513,93 @@ def test_launch_grpc_not_supported_version():
+
+    """If UserWarning & pythonnet are in the stderr output,
+    set warning to 0. Otherwise, set warning to 1"""
+    if "UserWarning" and "pythonnet" in stderr_output:
+        warning = 0
+    else:
+        warning = 1
+
+    # Assert the warning message did not appear for the remote session
+    assert 1 == warning
```
```suggestion
    warning = True if "UserWarning" and "pythonnet" in stderr_output else False

    # Assert the warning message did not appear for the remote session
    assert not warning
```

RobPasMue commented on tests/test_mechanical.py:
```diff
@@ -510,6 +513,93 @@ def test_launch_grpc_not_supported_version():
+    check_warning.wait()
+    stderr_output = check_warning.stderr.read().decode()
+
+    """If UserWarning & pythonnet are in the stderr output,
+    set warning to 0. Otherwise, set warning to 1"""
```
Better use in-line comments
```suggestion
   # If UserWarning & pythonnet are in the stderr output, set warning to True.
   # Otherwise, set warning to False
```

RobPasMue commented on tests/test_mechanical.py:
```diff
@@ -510,6 +513,93 @@ def test_launch_grpc_not_supported_version():
+    check_warning.wait()
+    stderr_output = check_warning.stderr.read().decode()
+
+    """If UserWarning & pythonnet are in the stderr output,
+    set warning to 0. Otherwise, set warning to 1"""
```
```suggestion
   # If UserWarning & pythonnet are in the stderr output, set warning to True.
   # Otherwise, set warning to False
```

RobPasMue commented on tests/embedding/test_app.py:
```diff
@@ -52,3 +54,67 @@ def test_app_version(embedded_app):
+    check_warning.wait()
+    stderr_output = check_warning.stderr.read().decode()
+
+    """If UserWarning & pythonnet are in the stderr output,
+    set warning to 1. Otherwise, set warning to 0"""
```
```suggestion
    # If UserWarning & pythonnet are in the stderr output, set warning to True.
    # Otherwise, set warning to False
```

RobPasMue reviewed (approved):
Tests are looking good! Nicely done @klmcadams. I left some suggestions/minor improvements. I'll leave it up to @koubaa to approve and merge, but for me, this PR is looking good.

koubaa reviewed (approved):
This implementation of a pythonnet check is fabulous - thank you for your contribution!

koubaa commented:
@klmcadams if you're willing to take any more I could suggest some
